### PR TITLE
[SR-1813] Fix archetype access path binding

### PR DESCRIPTION
**Pull request by CodaFi**
#### What's in this pull request?

Fixes a regression from Swift 2.2 where the configuration reported in SR-1813 would cause the compiler to forget to setup archetype access paths which would blow up witness table gen.
#### Resolved bug number: ([SR-1813](https://bugs.swift.org/browse/SR-1813))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by jrose-apple** [comment]
Probably best reviewed by @slavapestov or @rjmccall.

File: lib/IRGen/GenProto.cpp [hunk 1]
```diff
@@ -1669,35 +1669,44 @@ bool irgen::hasPolymorphicParameters(CanSILFunctionType ty) {
...
+  if (!srcBaseArchetype) return;
+
+  IGF.addArchetypeAccessPath(destArchetype,
+                             {srcBaseArchetype, association});
 }
 
-void EmitPolymorphicParameters::bindArchetypeAccessPaths() {
-  if (!Generics) return;
+static void bindArchetypeAccessPaths(IRGenFunction &IGF,
+                                     ArrayRef<Requirement> Generics,
```
Comment (rjmccall) [inline lib/IRGen/GenProto.cpp]: Please take a GenericSignature instead of an array of Requirements.

**Comment by rjmccall** [comment]
Generally looks okay, but:

- The diff is strange; I'm really not sure why github's diff is removing and re-adding EmitPolymorphicParameters::emit.  It might just be poor tooling, but it's odd enough to make me wonder whether you've accidentally changed line terminators or something.

- Try to always pass around generic signatures as GenericSignatures rather than an independent collection of requirements.

**Comment by CodaFi** [comment]
> The diff is strange;

I did have to shuffle a non-trivial amount of code around.

**Comment by rjmccall** [comment]
Yeah, maybe it's just that the function is small.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by CodaFi** [comment]
@rjmccall Mind giving it another go?

**Comment by rjmccall** [comment]
LGTM.

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@rjmccall ping

**Comment by harlanhaskins** [comment]
@rjmccall please ping

**Comment by banjun** [comment]
Thank you @CodaFi 👍
I'm reporter of the original issue and have confirmed that your commit resolve the issue.

However, I found another crash on this commit but I cannot tell whether this is same issue or not. (around `emitArchetypeWitnessTableRef`).
Should I submit another report?

The new case is below:

```
protocol P1 {
    associatedtype A1: Equatable
}

protocol P2 {
    associatedtype A2: Equatable
}

class C1<T, U: P1 where U.A1 == T>: P2 {
    typealias A2 = T
}
```

will result in:

> no relation found that declares conformance to target

```
6  swift                    0x0000000110b8da27 llvm::llvm_unreachable_internal(char const*, char const*, unsigned int) + 471
7  swift                    0x000000010df9ff00 llvm::Value* llvm::function_ref<llvm::Value* (unsigned int)>::callback_fn<swift::irgen::emitArchetypeWitnessTableRef(swift::irgen::IRGenFunction&, swift::CanTypeWrapper<swift::ArchetypeType>, swift::ProtocolDecl*)::$_0>(long, unsigned int) + 368
8  swift                    0x000000010e06e921 swift::irgen::emitImpliedWitnessTableRef(swift::irgen::IRGenFunction&, llvm::ArrayRef<swift::irgen::ProtocolEntry>, swift::ProtocolDecl*, llvm::function_ref<llvm::Value* (unsigned int)> const&) + 273
9  swift                    0x000000010df9d8c4 swift::irgen::emitArchetypeWitnessTableRef(swift::irgen::IRGenFunction&, swift::CanTypeWrapper<swift::ArchetypeType>, swift::ProtocolDecl*) + 164
```

Moreover, removing 'protocol P2' and using 'P1' instead of 'P2' still reproduce same error message and make the case shortest example. (I think using P1 and P2 is more general case).

**Comment by CodaFi** [comment]
Yes, please submit another SR to bugs.swift.org.  This should have shown up as a diagnostic and not survived to SILGen.

**Comment by banjun** [comment]
I submitted the other case to [SR-1939](https://bugs.swift.org/browse/SR-1939). Thanks!

**Pull request merged**